superbench/benchmarks/micro_benchmarks/dist_inference_cpp/dist_inference.cu

```diff
@@ -182,10 +191,91 @@ void InitializeABCDEF(std::vector<cublasLtHalf> &ha, int64_t size_a, std::vector
+    const int kNumWarmups = 10;
+    const int kNumTestRuns = 100;
+    *ret_algo_time_in_ms = -1;
+    // Benchmark all algorithms in given shape
+    CHECK_CUDA_ERROR(cudaEventCreate(&start_event));
+    CHECK_CUDA_ERROR(cudaEventCreate(&end_event));
+    // Partition work evenly into different ranks
+    for (size_t algo_idx = rank; algo_idx < gemm_heuristics.size(); algo_idx += num_ranks) {
+        auto &algo = gemm_heuristics[algo_idx].algo;
+        size_t ret_workspace_size = 0;
+        auto status = hipblaslt_ext::matmulIsAlgoSupported(handle, matmul, &alpha, matA, matB, &beta, matC, matD, algo,
+                                                           ret_workspace_size);
+        if (status != HIPBLAS_STATUS_SUCCESS || ret_workspace_size >= workspace_size) {
+            continue;
+        }
+        for (int i = 0; i < kNumWarmups; i++) {
+            CHECK_CUBLASLT_ERROR(hipblasLtMatmul(handle, matmul, &alpha, da, matA, db, matB, &beta, dc, matC, dd, matD,
+                                                 &algo, d_workspace, workspace_size, stream));
+        }
+        CHECK_CUDA_ERROR(cudaEventRecord(start_event, stream));
+        for (int i = 0; i < kNumTestRuns; i++) {
+            CHECK_CUBLASLT_ERROR(hipblasLtMatmul(handle, matmul, &alpha, da, matA, db, matB, &beta, dc, matC, dd, matD,
+                                                 &algo, d_workspace, workspace_size, stream));
+        }
+        CHECK_CUDA_ERROR(cudaEventRecord(end_event, stream));
+        CHECK_CUDA_ERROR(cudaStreamSynchronize(stream));
+        float time_in_ms = 0;
+        CHECK_CUDA_ERROR(cudaEventElapsedTime(&time_in_ms, start_event, end_event));
+        time_in_ms /= kNumTestRuns;
+        if (*ret_algo_time_in_ms < 0 || time_in_ms < *ret_algo_time_in_ms) {
+            *ret_algo = algo;
+            *ret_algo_time_in_ms = time_in_ms;
+        }
+    }
+    CHECK_CUDA_ERROR(cudaEventDestroy(start_event));
+    CHECK_CUDA_ERROR(cudaEventDestroy(end_event));
+}
+
+// Select global best GEMM algorithms across ranks. Write global_algo if something is found.
+void TuneHipblasLtGemmGlobal(int num_ranks, const hipblasLtMatmulAlgo_t &local_algo, float local_time_in_ms,
+                             hipblasLtMatmulAlgo_t *global_algo) {
+    std::vector<hipblasLtMatmulAlgo_t> coll_algo(num_ranks);
+    std::vector<float> coll_time_in_ms(num_ranks);
+    MPI_Allgather(&local_algo, sizeof(local_algo), MPI_BYTE, coll_algo.data(), sizeof(local_algo), MPI_BYTE,
+                  MPI_COMM_WORLD);
+    MPI_Allgather(&local_time_in_ms, sizeof(local_time_in_ms), MPI_BYTE, coll_time_in_ms.data(),
+                  sizeof(local_time_in_ms), MPI_BYTE, MPI_COMM_WORLD);
+    float min_time_in_ms = -1;
+    for (int i = 0; i < num_ranks; i++) {
+        if (coll_time_in_ms[i] >= 0 && (min_time_in_ms < 0 || coll_time_in_ms[i] < min_time_in_ms)) {
+            min_time_in_ms = coll_time_in_ms[i];
+            *global_algo = coll_algo[i];
+        }
+    }
+}
+#endif
```

Comment:
Therefore, it's not a library. Just a piece of code?

Reply:
TunableOp is a design that is implemented in PyTorch, onnx runtime, etc. It's not included in standalone library. Implementation in PyTorch for ref: https://github.com/pytorch/pytorch/commit/0e6eee3c898c293efc3f172180e9f4d79cc0b13f